eyetracking/bids_scripts/utils.py

```diff
@@ -0,0 +1,372 @@
+    gaze: np.array,
+    onset_time: float,
+    conf_thresh: float=0.9,
+    distance: bool=True,
+) -> tuple:
+    '''
+    Realign gaze timestamps based on the task & eyetracker onset/offset.
+    Export new list of gaze dictionaries
+    (w task-aligned time stamps) and other metrics needed to perform
+    drift correction
+    '''
+    # all gaze values (unfiltered)
+    reset_gaze_list = []
+    all_x = []
+    all_y = []
+    all_times = []
+    all_conf = []
+
+    """
+    Normalized (proportion of screen) distance between above-threshold gaze
+    and fixation point(if distance=True) or normalized position
+    (if distance=False).
+    """
+    (x_diff, y_diff) = [0.5, 0.5] if distance else [0.0, 0.0]
+
+    clean_x = []
+    clean_y = []
+    clean_times = []
+    clean_conf = []
+
+    for gz_pt in gaze:
+        timestp = gz_pt['timestamp'] - onset_time
+        # exclude gaze & pupil gathered before task onset
+        if timestp > 0.0:
+            x_norm, y_norm = gz_pt['norm_pos']
+            cfd = gz_pt['confidence']
+
+            all_x.append(x_norm)
+            all_y.append(y_norm)
+            all_times.append(timestp)
+            all_conf.append(cfd)
+
+            gz_pt['reset_time'] = timestp
+            reset_gaze_list.append(gz_pt)
+
+            if cfd > conf_thresh:
+                clean_x.append(x_norm - x_diff)
+                clean_y.append(y_norm - y_diff)
+                clean_conf.append(cfd)
+                clean_times.append(timestp)
+
+    return reset_gaze_list, (all_x, all_y, all_times, all_conf), (clean_x, clean_y, clean_times, clean_conf)
+
+
+def apply_poly(
+    ref_times: list,
+    distances: list,
+    degree: int,
+    all_times: np.array,
+    anchors: list = [150, 150],
+) -> np.array:
+    '''
+    Fit polynomial to a distribution, then export points along that polynomial curve
+    that correspond to specific gaze time stamps
+
+    The very begining and end of distribution are excluded for stability
+    e.g., participants sometimes look off screen at movie onset & offset,
+    while deepgaze is biased to the center when shown a black screen.
+    This really throws off polynomials
+    '''
+    if degree == 1:
+        p1, p0 = np.polyfit(ref_times[anchors[0]:-anchors[1]], distances[anchors[0]:-anchors[1]], 1)
+        p_of_all = p1*(all_times) + p0
+
+    elif degree == 2:
+        p2, p1, p0 = np.polyfit(ref_times[anchors[0]:-anchors[1]], distances[anchors[0]:-anchors[1]], 2)
+        p_of_all = p2*(all_times**2) + p1*(all_times) + p0
+
+    elif degree == 3:
+        p3, p2, p1, p0 = np.polyfit(ref_times[anchors[0]:-anchors[1]], distances[anchors[0]:-anchors[1]], 3)
+        p_of_all = p3*(all_times**3) + p2*(all_times**2) + p1*(all_times) + p0
+
+    elif degree == 4:
+        p4, p3, p2, p1, p0 = np.polyfit(ref_times[anchors[0]:-anchors[1]], distances[anchors[0]:-anchors[1]], 4)
+        p_of_all = p4*(all_times**4) + p3*(all_times**3) + p2*(all_times**2) + p1*(all_times) + p0
```

Comment:
There always exist a function to do things that common 😜 .
```suggestion
    ps = np.polyfit(ref_times[anchors[0]:-anchors[1]], distances[anchors[0]:-anchors[1]], degree)
    p_of_all = np.polyval(ps, all_times)
```